### ISSUE TITLE
Mypy Self Checking Should Enable --warn-unreachable

### ISSUE THREAD

**Issue by davidhalter** [python/mypy#18079]
Is it intentional that `--warn-unreachable` is not used while self checking in the CI? (with mypy_self_check.ini)

I have recently realized that this is not turned on and it feels a bit strange. There are about 39 errors when running with --warn-unreachable and it brings up a few questionable things like https://github.com/python/mypy/blob/master/mypy/test/testpep561.py#L180.

@sobolevn told me to create this issue in https://github.com/python/mypy/pull/18076#issuecomment-2449506357

**Comment by sobolevn** [comment]
Please, feel free to send a PR.

**Comment by AlexWaygood** [comment]
I tried this in the past and found that it caused too many false-positive errors in the self-check at that point to make it feasible. For example, branches under `if sys.platform == "win32"` tests were thought to be unreachable if checking with `--platform=linux` (strictly speaking that's true, but it's very annoying and not at all useful to have that cause the check to fail).

The reachability heuristics mypy uses may have improved since then, though; it was a while ago that I tried it.

**Comment by davidhalter** [comment]
> For example, branches under if sys.platform == "win32" tests were thought to be unreachable if checking with --platform=linux

I don't think this is the case anymore. I also haven't had any problems with it in my personal projects and going randomly through 8 of the 39 issues raised above, Mypy seems to be correct about all of them except maybe one that is related to `mypy_extensions.trait`. I'm not sure I have the time right now, but I really believe that this would be worth it.

**Comment by JelleZijlstra** [comment]
I think turning this option on for mypy itself would be a good way to dogfood it and force ourselves to make sure it is reliable.

**Comment by AlexWaygood** [comment]
I _absolutely_ agree that this would be worth it if it is possible! I wasn't trying to dissuade anybody, was just giving context on something I previously tried a while back :-)

And, it may well be that the check has improved significantly since I tried it last!

**Comment by A5rocks** [comment]
I started doing this (as a blocker for another mypy change) and the first 3 errors I had to fix were just https://github.com/python/mypy/issues/10773... So I think that blocks this. Things like this:

```py
def parse_gray_color(cup: bytes) -> str:
    """Reproduce a gray color in ANSI escape sequence"""
    if sys.platform == "win32":
        assert False, "curses is not available on Windows"
    set_color = "".join([cup[:-1].decode(), "m"])  # E: Statement is unreachable
    gray = curses.tparm(set_color.encode("utf-8"), 1, 9).decode()
    return gray
```

(obviously this can be *made* reachable but IMO it's better to avoid that churn)
